### ISSUE TITLE
Add support for W3C conform multiple CORS origins

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#3745]
# This is a Feature Proposal

## Description

This is a follow-up issue on https://github.com/serverless/serverless/issues/2476 since we've implemented the `origin` config parameter to make it possible to define one origin per event definition (as API Gateway proposes it).

However sometimes users might want to define multiple origins. We should discuss this feature in more-depth in this issue since it might introduce some implementation challenges do to lack of direct support via CloudFormation.

More discussion about this can be found [here](https://github.com/serverless/serverless/issues/2476#issuecomment-304867269).

/cc @alexcasalboni @HyperBrain @jokeyrhyme

**Comment by gaastonsr** [comment]
This is so needed.

**Comment by a-h** [comment]
Here's my thoughts on it...

## Example scenario

I have an API which I want two websites to be able to use so I've used the Serverless Framework's `cors: true` setting.

This could be exploited by an attacker. To do this, an attack could exploit a popular site and place Javascript code which accesses my API repeatedly on it.

Since `cors: true` sets the `access-control-allow-origin` header to `*`, Web browsers will allow HTTP requests originating from any web page and would therefore not prevent the HTTP requests, so my API would be flooded with requests from many different locations, increasing my costs, and possibly causing an outage.

## Ways to support multiple allowed origins instead of just a wildcard

To support multiple allowed origins instead of the wildcard, the CORS preflight check must return a single origin (comma separated etc. is not supported by browsers), so if the `origin: a.com` header is received and the origin should be allowed to access the API, it should respond with `access-control-allow-origin: a.com`.

In code, it's easy enough - read the `origin` header value, check against the list of allowed origins and set the `access-control-allow-origin` header to the value of the input `origin` header if it's in the list. 

### 1: Use a Lambda

Any non OPTIONS response needs to include this anyway, so one implementation could be to use a Lambda to do this - i.e. Serverless would create a Lambda just to do this work. This Lambda wouldn't need to have the same IAM permissions as the actual application Lambdas, or be in a VPC etc. which would help to reduce the impact of requests coming in.

### 2: Build on the API Gateway Mock

API Gateway is limited in what it can do. While body mapping templates can contain logic, there's no support I know of to use logic to set headers, but one approach could be to use (abuse?) the interplay between the Mock integration and Integration Response.

Setting `cors: true` creates an "OPTIONS" method against the resource in API Gateway. This OPTIONS method is setup against a Mock integration. The Mock integration returns the hard-coded response `{statusCode:200}` which is passed on to the Integration Response. The integration response returns a hard coded set of headers on receipt of that `statusCode=200`, resulting in setting the `access-control-allow-origin` header to `*`.

However, the Mock integration has ability to use a body mapping template, so it can read the incoming `origin` header, then output a different HTTP status code depending on whether the origin was in the list of allowed origins or not (e.g. a.com = 201, b.com = 202, notallowed.com = 100). The Integration Response could then respond with a different `access-control-allow-origin` value depending on what the Mock integration output.

<img width="835" alt="screen shot 2018-03-07 at 13 58 34" src="https://user-images.githubusercontent.com/1029947/37096384-59df27d8-2210-11e8-9433-4d24b731f626.png">

This would potentially avoid the need to run a lambda to deliver the functionality.

**Comment by yibinlin** [comment]
+1 really needed feature for most testing version of the website and some production websites too. I am not sure if a separate lambda function will bring some performance penalties? (Or if the answer is yes, is the performance penalty bearable in common use cases?)

**Comment by jokeyrhyme** [comment]
I hinted at an alternative approach here: https://github.com/serverless/serverless/issues/2476#issuecomment-304985035

An example implementation with more detail:
1. Serverless reads the configuration for the endpoint, notes the user-desired handler / entrypoint and that CORS with multiple origins is configured
2. Serverless generates a wrapper function:
  - the function includes the multiple origins, and the logic to compare against the incoming origin
  - the function includes a call to the user-desired handler as appropriate
  - the user-desired handler should be passed all information as normal, should not know it was wrapped
3. Serverless configures Lambda (etc) so that the wrapper will be called instead of the desired handler

This is very similar to @a-h 's Option 1 "Use a Lambda", except the extra code is included in with each deployed Lambda instead of in a separate Lambda
Theoretically this might be lower latency that having one Lambda call another Lambda
The downside is the wrapper function interactions add complexity, and may even need to be language specific

**Comment by yibinlin** [comment]
Just would like to check if there are any update on this. Thank you!

**Comment by aldentaylor** [comment]
Any update on this?

**Comment by Borduhh** [comment]
Any update on this?

**Comment by ChrisSargent** [comment]
We seem to be having this issue as well: https://github.com/serverless/serverless/issues/3745#issuecomment-494433968

When setting up using multiple CORS origins, we always get the error above when a pre-flight request hits API-GW. It works fine if there is no pre-flight and the logic is handled by the lambda function itself.

**Comment by ChrisSargent** [comment]
Is it maybe something to do with this code here: https://github.com/serverless/serverless/blob/e32f46ee8c0677072d9511228d1ea1db60bb8c36/lib/plugins/aws/package/compile/events/apiGateway/lib/cors.js#L66

It seems to just send the origins. And then in the template, we just get something like:

```
"method.response.header.Access-Control-Allow-Origin": "'https://dev.app.co,https://dashboard-dev.app.co,https://local.app.co:3000'"
```

Though I'm not certain if there's some run time processing going on because that seems to be setting the mock responses in the template with `"Type": "MOCK",`

**Comment by jkaldon-rebellabs** [comment]
We too could really use this functionality.  Admittedly, it feels like something that the ApiGateway should be able to handle.

**Comment by Aristarhys** [comment]
@ChrisSargent could you please provide a snippet of how to setup multiple origins CORS lambda can handle logic without pre-flight? You mean implementing OPTIONS response yourself?

**Comment by Aristarhys** [comment]
I have similar issue and i wonder if https://github.com/serverless/serverless/blob/f93b27bf684d9a14b1e67ec554a7719ca3628135/lib/plugins/aws/package/compile/events/apiGateway/lib/method/responses.js#L15 is source of the problem just like generated fishy looking templates with origin set as it is on serverless.yml

**Comment by ChrisSargent** [comment]
> @ChrisSargent could you please provide a snippet of how to setup multiple origins CORS lambda can handle logic without pre-flight? You mean implementing OPTIONS response yourself?

Hi @Aristarhys - I think we took out all the cors config from serverless.yml (and therefore APIGW). We then had a 'catch-all' endpoint running an Express server (from some legacy code as we were in the middle of migrating to serverless) which was supplying the correct CORS headers.

See below for some notes on this...


# Cors Errors

Note: to run Chrome whilst ignoring CORS errors, use `open -a Google\ Chrome --args --disable-web-security --user-data-dir`

1. Added a `cors:` option to every endpoint as detailed here: https://serverless.com/blog/cors-api-gateway-survival-guide/#cors-preflight-requests - specifically > Simply add cors: true to each endpoint in your serverless.yml:

# Failed Experiments:

- cors: true
  Result

```
no access control header
```

- cors: '\*'
- creds: true
- Result

```
Access to XMLHttpRequest at 'https://api-dev.jemjar.crownd.co/card-linking/cards/454931' from origin 'https://local.jemjar.crownd.co:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.
```

- cors: '\*'
- creds: false
- Result

```
Access to XMLHttpRequest at 'https://api-dev.jemjar.crownd.co/user/consumer' from origin 'https://local.jemjar.crownd.co:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.
```

- cors: CONFIG
- creds: false
- Result

```
Access to XMLHttpRequest at 'https://api-dev.jemjar.crownd.co/user/consumer' from origin 'https://local.jemjar.crownd.co:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The 'Access-Control-Allow-Origin' header contains multiple values 'https://dev.jemjar.crownd.co,https://dashboard-dev.jemjar.crownd.co,https://local.jemjar.crownd.co:3000', but only one is allowed.
```

- cors: CONFIG
- creds: true
- Result

```
Access to XMLHttpRequest at 'https://api-dev.jemjar.crownd.co/user/consumer' from origin 'https://local.jemjar.crownd.co:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The 'Access-Control-Allow-Origin' header contains multiple values 'https://dev.jemjar.crownd.co,https://dashboard-dev.jemjar.crownd.co,https://local.jemjar.crownd.co:3000', but only one is allowed.
```

# Successful Experiment!

- Completely removed CORS settings from the `serverless.yml` file, everything works as expected. All endpoints work and respond with the correct headers for both the pre-flight and main request. Calls from non-listed URLs get blocked as expected.

# Observations

- This issue is definitely only happening when we have a preflight request. Supposedly on a pre-flight (OPTIONS) request, the browser should _not_ send authentication, including cookies, see: https://stackoverflow.com/questions/41760128/cookies-not-sent-on-options-requests, and: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Preflighted_requests
- API-GW handles pre-flight requests
- Previously, calls to the old API endpoint were working okay, but it did not have the `cors` option specified in `serverlesss.yml`. After adding this option, we started getting the multiple CORS origin errors.
- According to Firefox, the cookies are not being included by the browser / Axios in the OPTIONS request - which is correct

# Unrelated but Worth Noting

You need to set up CORS for error responses, see: https://serverless-stack.com/chapters/handle-api-gateway-cors-errors.html

# Useful Links
- https://github.com/serverless/serverless/issues/3745
- https://github.com/serverless/serverless/issues/2476
- https://github.com/serverless/serverless/blob/e32f46ee8c0677072d9511228d1ea1db60bb8c36/lib/plugins/aws/package/compile/events/apiGateway/lib/cors.js#L66
- https://github.com/shekhargulati/hands-on-serverless-guide/blob/master/01-aws-lambda-serverless-framework/04-enable-cors-for-your-rest-api.md
- https://serverless.com/framework/docs/providers/aws/events/apigateway#enabling-cors Specifically, "If you want to use CORS with the lambda-proxy integration, remember to include the Access-Control-Allow-* headers in your headers object, like this:"
- https://stackoverflow.com/questions/43757019/how-to-add-cors-header-to-aws-api-gateway-response-with-lambda-proxy-integration
- https://forum.serverless.com/t/access-control-allow-methods-collects-methods-from-multiple-paths/4368

**Comment by medikoo** [comment]
Wrapping lambdas is out of scope in Serverless Framework. At least at this point we do not consider solving any issues that way (as its invasive and brings tons of implications).

_(it's only our dashboard support which requires wrapping lambdas to inject our SDK, but that's served by separate plugin in other repository)_

Therefore if there's no easy way to fix this on CloudFormation grounds, we need to coin it as provider limitation, and have  to keep the issue unresolved until it's the case, and recommend above [workaround](https://github.com/serverless/serverless/issues/3745#issuecomment-523843695) as posted by  @ChrisSargent 

Have any one tried to contact an AWS support about that? I wonder what approach is recommended by them.

**Comment by yehonadav** [comment]
nobody wants to fix this?
kinda sucks to create preflight for every endpoint just so you can allow more then 1 origin 💩

**Comment by rodrigomf24** [comment]
If you are bashing your head against this mind-numbing problem, here is what worked for me, none of the "solutions" I encountered fixed the problem for me, but this one did:

https://blog.rob.uk.com/2019/01/26/Allowing-multiple-CORS-origins-with-AWS-API-Gateway/

Yes, it sucks to have a definition for every lambda, but it is the only solution that has successfully worked for me:

This is my implementation using typescript:

```
import {
  APIGatewayProxyCallback,
  APIGatewayProxyEventV2,
  Context,
} from 'aws-lambda';

const ALLOWED_ORIGINS_REGEX = [
  new RegExp(/https:\/\/(.*).somedomain.com/),
  new RegExp(/https:\/\/(.*).anotherdomain.com/),
]

export const getCorsOrigin = (incomingOrigin?: string) => {
  if (!incomingOrigin) {
    return null;
  }

  const isAllowedOrigin = ALLOWED_ORIGINS_REGEX.find((whitelistedOrigin) => {
    if (whitelistedOrigin instanceof RegExp) {
      return whitelistedOrigin.test(incomingOrigin);
    }

    return incomingOrigin === whitelistedOrigin;
  });

  if (incomingOrigin && isAllowedOrigin) {
    return incomingOrigin;
  }

  return null;
};

export const resolver = async (
  event: APIGatewayProxyEventV2,
  _context: Context,
  callback: APIGatewayProxyCallback
) => {
  let requestOrigin = event.headers["Origin"];

  if (!requestOrigin || !requestOrigin.length) {
    requestOrigin = event.headers["origin"];
  }

  const allowedOrigin = evaluateRequestOrigin(requestOrigin);

  const nextHeaders: Record<string, string | boolean | number> = {};

  if (allowedOrigin) {
    nextHeaders["Access-Control-Allow-Credentials"] = true;
    nextHeaders["Access-Control-Allow-Origin"] = allowedOrigin;
    nextHeaders["Access-Control-Allow-Headers"] =
      "Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token";
    nextHeaders["Access-Control-Allow-Methods"] = "GET, POST, OPTIONS";
  }

  callback(null, {
    statusCode: 200,
    headers: nextHeaders,
    body: "",
  });
};

  ```
  
😮‍💨  I hope this helps anyone that stumbles into this thread...
